RevenueCatUI/CustomerCenter/ViewModels/CustomerCenterViewModel.swift

```diff
@@ -0,0 +1,99 @@
+//
+//  Copyright RevenueCat Inc. All Rights Reserved.
+//
+//  Licensed under the MIT License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      https://opensource.org/licenses/MIT
+//
+//  CustomerCenterViewModel.swift
+//
+//
+//  Created by Cesar de la Vega on 27/5/24.
+//
+
+import Foundation
+import RevenueCat
+
+@available(iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0, *)
+@available(macOS, unavailable)
+@available(tvOS, unavailable)
+@available(watchOS, unavailable)
+@MainActor class CustomerCenterViewModel: ObservableObject {
+
+    typealias CustomerInfoFetcher = @Sendable () async throws -> CustomerInfo
+
+    @Published
+    var hasSubscriptions: Bool = false
+    @Published
+    var subscriptionsAreFromApple: Bool = false
+    @Published
+    var state: CustomerCenterViewState {
+        didSet {
+            if case let .error(stateError) = state {
+                self.error = stateError
+            }
+        }
+    }
+
+    var isLoaded: Bool {
+        return state != .notLoaded
+    }
+
+    private var customerInfoFetcher: CustomerInfoFetcher
+
+    private var error: Error?
+
+    convenience init() {
+        self.init(customerInfoFetcher: {
+            guard Purchases.isConfigured else {
+                throw PaywallError.purchasesNotConfigured
+            }
+
+            return try await Purchases.shared.customerInfo()
+        })
+    }
+
+    // @PublicForExternalTesting
+    init(customerInfoFetcher: @escaping CustomerInfoFetcher) {
+        self.state = .notLoaded
+        self.customerInfoFetcher = customerInfoFetcher
+    }
+
+    // @PublicForExternalTesting
+    init(hasSubscriptions: Bool = false, areSubscriptionsFromApple: Bool = false) {
+        self.hasSubscriptions = hasSubscriptions
+        self.subscriptionsAreFromApple = areSubscriptionsFromApple
+        self.customerInfoFetcher = {
+            guard Purchases.isConfigured else {
+                throw PaywallError.purchasesNotConfigured
+            }
+
+            return try await Purchases.shared.customerInfo()
+        }
+        self.state = .success
+    }
+
+    func loadHasSubscriptions() async {
+        do {
+            // swiftlint:disable:next todo
+            // TODO: support non-consumables
+            let customerInfo = try await self.customerInfoFetcher()
+            let hasSubscriptions = customerInfo.activeSubscriptions.count > 0
+
+            let subscriptionsAreFromApple = customerInfo.entitlements.active.first(where: {
+                $0.value.store == .appStore || $0.value.store == .macAppStore
+            }).map { entitlement in
+                customerInfo.activeSubscriptions.contains(entitlement.value.productIdentifier)
+            } ?? false
```

Comment:
Hmmm, yes I think if the first active entitlement comes from a consumable, but the second active entitlement is from a subscription, this will be false. I think I need to combine this. Good catch

Reply:
I think I addressed it in [`9c89313` (#3919)](https://github.com/RevenueCat/purchases-ios/pull/3919/commits/9c89313c18df5b2db0dabed33de6c8855268bca6)